### PR TITLE
🎉 New Source: Apple Search Ads [low-code CDK]

### LINKED ISSUE #2982: New source: Apple Search Ads Reports
## Tell us about the new integration you’d like to have
https://searchads.apple.com/advanced/

## Describe the alternative you are considering or using

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368161100890) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by rach-r**
## What
This PR aims to add a new connector fetching data from the [Apple Search Ads API](https://developer.apple.com/documentation/apple_search_ads). 

~~Blocked by [#15268](https://github.com/airbytehq/airbyte/issues/15628).~~
Closes airbytehq/airbyte#2982.

## How

It's based on the [Airbyte low-code CDK.](https://docs.airbyte.com/connector-development/config-based/low-code-cdk-overview/)

It offers two kind of streams:
- Base streams, which contains attributes for entities (campaigns, adgroups, keywords)
- Report streams, which contains statistics about the performance of these entities (# of clic, amount spent etc)
  - The API offers `HOURLY`, `DAILY`, `WEEKLY`, `MONTHLY` granularity
  - But this connector only offer report with `DAILY` granularity at this time

~~Before being merged, it needs [this issue](https://github.com/airbytehq/airbyte/issues/15628) to be fixed. Why?~~
~~Because the [Apple Search Ads has a 30d attribution window](https://developer.apple.com/documentation/iad/setting_up_apple_search_ads_attribution), so we need:~~
~~- The report streams to have Incremental Dedup sync~~
~~- A `lookback_window` set to `30d`~~
~~But in the low-code CDK, the `lookback_window` is currently applied at the `start_date` instead of the `cursor_field`, which defeats its purpose (only applied on first sync and never on the following syncs, see Issue for details).~~

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### CODE REVIEW

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (rach-r) [inline airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:105]: The `lookback_window` is critical here because this API fetches marketing data with an [attribution window of 30 days](https://developer.apple.com/documentation/iad/setting_up_apple_search_ads_attribution), ie any data < 30 days needs to be refreshed with an Incremental Dedup sync.

However, the `lookback_window` is currently bugged as it is applied on the `start_date` instead of the `cursor_field`: issue #15628

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (rach-r) [inline airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:99]: I would like to do something like `max(config['start_date'], day_delta(-90))` as the daily reports are limited to 90 days in the past and trying to read beyond will trigger an error. 

But the `day_delta` [macro](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py#L86) is a datetime while I need a date. Am I allowed to add a macro? I don't want to clutter the macros with too specific cases.

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (rach-r) [inline airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:161]: As nested cursors are not supported yet for `stream_cursor_field` (and even for `primary_key` for normalization?) I am adding them as new fields at the root of the each stream (thanks @brianjlai for the tip 👌)

**Comment by rach-r** [comment]
I am not able to run acceptance tests:
- `secrets/config.json` does exist and is at root of connector folder
- I built the image using docker `build . -t airbyte/apple-search-ads:dev`
- Then I'm running the tests successfully with:
```
docker run --rm airbyte/connector-name:dev spec
docker run --rm -v $(pwd)/secrets:/secrets airbyte/apple-search-ads:dev check --config /secrets/config.json
docker run --rm -v $(pwd)/secrets:/secrets airbyte/apple-search-ads:dev discover --config /secrets/config.json
docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/apple-search-ads:dev read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json
```
- Then I run `python -m pytest integration_tests -p integration_tests.acceptance`
- Everything is fine until 50%, when I get to the check --config part: 
```
FAILED ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection::test_check[inputs0] - docker.errors.ContainerError: Command 'check --config /data/tap_config.json'
```
- Then all following tests fail with `FileNotFoundError: [Errno 2] No such file or directory: '/data/tap_config.json'`

It looks like the `secrets/config.json` file is not correctly detected. I've found someone had the exact same issue [here](https://github.com/airbytehq/airbyte/pull/15880) and also [here](https://github.com/airbytehq/airbyte/pull/17452#issuecomment-1307348140) but I don't have details regarding how it's been solved.

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by marcosmarxm** [comment]
So sorry missing this @rach-r I'll take a look later today!

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:99]: You can use datetime and use formatting to date, what do you think?

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:105]: @girarda any ideas if it's possible to solve #15628

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/spec.yaml
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/spec.yaml:1]:
```suggestion
documentationUrl: https://docs.airbyte.io/integrations/sources/apple-search-ads
```

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
The connector implementation looks solid. There are some issues blocking some features. A few comments to improve the code.

@rach-r do you have a test account to share?

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (girarda) [reply airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:105]: We don't have an ETA on this issue yet.

@rach-r https://github.com/airbytehq/airbyte/issues/15628 is a good first issue if this is blocking you and you're interested in contributing. Your contribution would be more than welcome!

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (rach-r) [reply airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:99]: Yes I've tried that a few weeks ago but could not figure how to put it in the interpolated string 😅  

However I just discovered this is not needed because there's actually no limitation in the `start_date`: setting it >=90 days does not throw any error (and returns data) so I don't need to limit this date parameter 👍

**Comment by rach-r** [comment]
Hey @marcosmarxm, first thanks again for taking time to review this PR 🙏 

I've done the requested changes + removed the 90 days limitation on the `start_date` (there's actually no limitation).

Unfortunately I don't have any test account and from what I've seen Apple does not provide any sandbox to make testing easier for this API 😕 from what I've read [here](https://github.com/airbytehq/airbyte/issues/2982#issuecomment-866008589) there's maybe already a test account for Airbyte? (unless I've misunderstood the conversation)

File: airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml
Comment (rach-r) [reply airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml:107]: @girarda I can see the issue is solved now, fixing the connector in Incremental Dedup mode. Thank you and @maxi297 for your help 🙏

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

File: airbyte-integrations/connectors/source-apple-search-ads/acceptance-test-config.yml
Comment (alafanechere) [inline airbyte-integrations/connectors/source-apple-search-ads/acceptance-test-config.yml:26]:
```suggestion
        future_state:
          future_state_path: "integration_tests/abnormal_state.json"
```
There was a typo in the code generator (now fixed), please make this minor change.

**Review by alafanechere — CHANGES_REQUESTED** [review]
@rach-r could you please share the output of the acceptance tests after the fix I'm suggesting?
`./gradlew :airbyte-integrations:connectors:source-apple-search-ads:integrationTest`

File: airbyte-integrations/connectors/source-apple-search-ads/acceptance-test-config.yml
Comment (rach-r) [reply airbyte-integrations/connectors/source-apple-search-ads/acceptance-test-config.yml]: I'm finally able to run the tests! Thank you 🙌 
I'm doing some changes as the CDK evolved since the PR was opened, and I'll post the results soon

Comment (rach-r) [reply airbyte-integrations/connectors/source-apple-search-ads/acceptance-test-config.yml]: @alafanechere I've been able to run the tests on a 2days lookback window:
image

However I'm not able to run them in a 30days lookback window as it times out eventually (lots of data to retrieve on our company account) - even if the `start_date` is set to today. I could increase the timeout but I don't think the acceptance tests should run for days? 😬 I've mentioned this issue in the Slack channel [here](https://airbytehq.slack.com/archives/C027KKE4BCZ/p1674215189136759) but could not find any workaround yet

**Review by alafanechere — APPROVED** [review]
Thank you for sharing the test output @rach-r . I'll let @marcosmarxm decide if we merge or not. We shall not publish this connector if we can't run SAT on our side. We're currently creating a sandbox account (tracked [here](https://github.com/airbytehq/airbyte/issues/21192)).

**Comment by alafanechere** [comment]
cc @sh4sh @sajarin @natalyjazzviolin as @marcosmarxm is off.

**Comment by rach-r** [comment]
> Thank you for sharing the test output @rach-r . I'll let @marcosmarxm decide if we merge or not. We shall not publish this connector if we can't run SAT on our side. We're currently creating a sandbox account (tracked [here](https://github.com/airbytehq/airbyte/issues/21192)).

I totally understand. Thanks a lot for taking time to review the PR anyway, I've learned a lot in the process and it will help me for future connectors 🙏

**Comment by rach-r** [comment]
I see #21192 is closed 🎉 I am rebasing and updating the PR to fit latest CDK changes

**Comment by rach-r** [comment]
@marcosmarxm @alafanechere If you want to test the sandbox 😇 I updated the connector with latest CDK and integration tests are still passing (with 1 day of lookback window as I have too much data): 
image

**Comment by marcosmarxm** [comment]
/publish connector=connectors/source-apple-search-ads run-tests=false

**Review by marcosmarxm — APPROVED** [review]

**Comment by MrdjanPoletanovic** [comment]
Hi, is it possible to add a timezone option?
Apple Search Ads API returns data by UTC timezone, and it would be useful to have a timezone option because it is not possible to do any type of transformation due to DAILY aggregated data. I need America/Los Angeles timezone, and there is no way to ensure it with the current connector implementation.

**Pull request merged**
